### ISSUE TITLE
Style/StringLiterals incorrect auto-correction #2

### ISSUE THREAD

**Issue by pirj** [rubocop-hq/rubocop#9394]
Follow-up to https://github.com/rubocop-hq/rubocop/issues/9393

```yaml
# .rubocop.yml
Style/StringLiterals:
  Enabled: true
  EnforcedStyle: single_quotes
  ConsistentQuotesInMultiline: false
```

```ruby
class Foo
  def bar
    """
      Hello, #{baz}
      Bye 2385-0138 -013-08 1-048 1-2084 -10284-0182-0
    """
  end

  def baz
    'World'
  end
end
```

```
$ rubocop --only Style/StringLiterals
Inspecting 1 file
C

Offenses:

1.rb:3:5: C: [Correctable] Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols.
    """
    ^^
1.rb:6:6: C: [Correctable] Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols.
    """
     ^^
```

```
rubocop --auto-correct --only Style/StringLiterals
```

```diff
class Foo
  def bar
-    """
+    ''"
      Hello, #{baz}
      Bye 2385-0138 -013-08 1-048 1-2084 -10284-0182-0
-    """
+    "''
```

:scream: 

--------

## Expected behavior

Does not raise an offence when there's string interpolation in a multi-line string literal.

## Actual behavior

Raises an offence and auto-corrects.

## Steps to reproduce the problem

See above

## RuboCop version

```
$ rubocop -V
warning: parser/current is loading parser/ruby25, which recognizes
warning: 2.5.8-compliant syntax, but you are running 2.5.7.
warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
1.8.1 (using Parser 3.0.0.0, rubocop-ast 1.4.0, running on ruby 2.5.7 x86_64-linux)
  - rubocop-minitest 0.10.3
  - rubocop-performance 1.9.2
  - rubocop-rails 2.9.1
  - rubocop-rspec 2.1.0
```

**Comment by koic** [comment]
I think this is an expected auto-correction behavior.

It has three strings.

1. blank

```ruby
""
```


2. interpolation

```ruby
"
  Hello, #{baz}
  Bye 2385-0138 -013-08 1-048 1-2084 -10284-0182-0
"
```

3. blank

```ruby
""
```

So I think the non-interpolated (blank) strings 1 and 2 should be auto-corrected to single quotes.

**Comment by pirj** [comment]
Sorry for the confusion, I think I've copied over the autocorrected code from the previous ticket, #9393. Fixed.

The point here is that the cop doesn't seem to detect multi-line strings with interpolation properly.

I see your point, the AST is a:
```
(dstr
  (str "")
  (dstr
    (str "\n")
    (str "Hello, ")
    (begin
      (send nil :bar))
    (str "\n")
    (str "Bye 2385-0138 -013-08 1-048 1-2084 -10284-0182-0\n"))
  (str ""))
```
and this whole construct is a string literal.

If we'd "correct" it to use tripe single quotes instead, won't it lose its interpolation properties becoming
```
(dstr
  (str "")
  (dstr
    (str "\n")
    (str "Hello, \#{bar}\n")
    (str "Bye 2385-0138 -013-08 1-048 1-2084 -10284-0182-0\n"))
  (str ""))
```
instead?

The call to `Foo.new.bar` would also produce a completely different result.

From my point of view, if any of the consisting parts is a `dstr`, we shouldn't flip from double quotes to single quotes.
I would suggest this change to `lib/rubocop/cop/style/string_literals.rb`:
```diff
        def all_string_literals?(nodes)
-          nodes.all? { |n| n.str_type? || n.dstr_type? }
+          nodes.all?(&:str_type?)
        end
```
This breaks the only example that I find hard to agree to, except for the point of having consistent quoting:
```
      it "doesn't register offense for double quotes with interpolation" do
        expect_no_offenses(<<~'RUBY')
          "abc" \
          "def#{1}"
        RUBY
      end
```

The complete change to skip changing quoting for string literals with any `dstr`s in it:
```diff
───────────────────────────────────────────────────────────────────────────────
modified: lib/rubocop/cop/style/string_literals.rb
───────────────────────────────────────────────────────────────────────────────
@ lib/rubocop/cop/style/string_literals.rb:37 @ module RuboCop
        MSG_INCONSISTENT = 'Inconsistent quote style.'

        def on_dstr(node)
+          return if node.parent&.dstr_type?
          # Strings which are continued across multiple lines using \
          # are parsed as a `dstr` node with `str` children
          # If one part of that continued string contains interpolations,
@ lib/rubocop/cop/style/string_literals.rb:46 @ module RuboCop
          return if node.heredoc?

          children = node.children
-          return unless all_string_literals?(children)
+          ignore_node(node) && return if any_interpolation?(children)

          quote_styles = detect_quote_styles(node)

@ lib/rubocop/cop/style/string_literals.rb:71 @ module RuboCop
          end
        end

-        def all_string_literals?(nodes)
-          nodes.all? { |n| n.str_type? || n.dstr_type? }
+        def any_interpolation?(nodes)
+          nodes.any?(&:dstr_type?)
        end

        def detect_quote_styles(node)
@ lib/rubocop/cop/style/string_literals.rb:84 @ module RuboCop
          # is nil. The quote marks are in the parent node.
          return [node.loc.begin.source] if styles.all?(&:nil?)

-          styles.map(&:source).uniq
+          styles.compact.map(&:source).uniq
        end

        def message(_node)
```
And it breaks no specs.

Would you accept such a pull request?
I'll add specs for multi-line string literals with triple quotes.

**Comment by marcandre** [comment]
> If we'd "correct" it to use tripe single quotes instead

It doesn't. It corrects using two single and a double quote `'` + `'` + `"`. It's meaningless, but so are triple quotes.

TBH, using "triple quotes" is a mistake in itself. Many people are confused in thinking that there is such a thing, or that one needs this to have multiple line strings (other than heredocs), but that is simply incorrect.
Maybe a different cop could handle this.

**Comment by pirj** [comment]
I agree that triple-quotes is a huge mistake, as it behaves completely different from, say, bash.

However, it's a valid Ruby syntax, and even if that other cop that replaced triple-quotes with single quotes, or a `\`-connected string literals with an explicit `\n`, or with heredoc, there is still a possibility that this other cop will be disabled in the user configuration.

From my point of view, this means that StringLiteral has to account for this syntax, and at least don't perform such a weird auto-correction.

**Comment by marcandre** [comment]
> it's a valid Ruby syntax

Indeed. `''" string "''` is also valid Ruby syntax.

Personally my opinion is that `"hello" "world#{2+2}" "!"` should raise an offense, and also `"""world#{2+2}!"""`. Maybe the autocorrection can be improved though.

**Comment by pirj** [comment]
My bad. I was under the impression that `''"` will not do interpolation, and will add the double quote to be part of the string.
Sorry for the confusion.
